### PR TITLE
Fix toctou

### PR DESCRIPTION
**Pull request by basilnsindhu**
*Description of changes:*

The current `now()` API call is prone to TOCTOU (time-of-check to time-of-use) error. This adds an additional API call which accepts a callback to an execution that needs to be timestamped by the user and calculates the upper and lower bound based on that. 
```
pub struct TimingResult {
    /// Callback began executing no earlier than this time
    pub earliest_start : SystemTime,
    /// Callback finished executing no later than this time
    pub latest_finish : SystemTime,
    /// No less than this amount of time elapsed from when timing() was called
    /// to when it returned.
    pub min_execution_time : Duration,
    /// No more than this amount of time elapsed when the callback was invoked
    /// to when it returned.
    pub max_execution_time : Duration
}
```
It also returns maximum and minimum execution time of the timing API call. 

Overview of the execution steps taken by the API call
```
1. User calls the `timing(f)` API 
2. Collects the upper and lower bound at T1 
3. Executes the callback f
4. Collects the upper and lower bound at T2
5. Compares duration between T1 midpoint and T2 midpoint
6. Computes maximum and minimum duration by +/- clock frequency
7. Returns upper bound, lower bound, min execution, maximum execution
``` 

Sample output of running `cargo run --example timing /run/clockboundd/clockboundd.sock`
```
Earliest start time for the timing request: "2022-02-21 19:43:34.530780989"
Latest finish time for the timing request: "2022-02-21 19:43:34.552979011"
Minimum execution duration of timing request: 51µs
Maximum execution duration of timing request: 73µs
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: clock-bound-c/src/lib.rs [hunk 1]
```diff
@@ -339,6 +359,71 @@ impl ClockBoundClient {
...
+        let duration_finish = latest_start.checked_add(Duration::from_nanos(latest_time_elapsed_midpoint)).unwrap();
+
+        // Calculates duration between the two midpoints
+        let duration = duration_finish.duration_since(duration_start).unwrap();
+        let ppm = Duration::from_micros(DEFAULT_CLOCK_FREQUENCY);
+
+        Ok((TimingResult{
+            earliest_start,
+            latest_finish,
+            min_execution_time: duration.saturating_sub(ppm),
```
Comment (dfoxfranke) [inline clock-bound-c/src/lib.rs:423]: This math isn't correct. If adjust by ±1ppm, you want to be multiplying by 0.999999 and 1.000001. Instead you're in effect adding or subtracting one microsecond, regardless of estimated elapsed time.

File: clock-bound-c/src/lib.rs [hunk 2]
```diff
@@ -339,6 +359,71 @@ impl ClockBoundClient {
...
+        }
+        let mut response: [u8; 20] = [0; 20];
+        match self.socket.recv(&mut response) {
+            Err(e) => return Err((ClockBoundCError::ReceiveMessageError(e), Ok(callback))),
+            _ => {}
+        }
+        let response_latest_start = NetworkEndian::read_u64(&response[4..12]);
+        let response_latest_finish = NetworkEndian::read_u64(&response[12..20]);
+
+        // Calculate elapsed time between early and latest timestamps
```
Comment (dfoxfranke) [inline clock-bound-c/src/lib.rs:404]: You get the midpoint by *adding* and dividing by 2, not by subtracting.

File: clock-bound-c/src/lib.rs [hunk 3]
```diff
@@ -95,7 +95,6 @@ pub struct ResponseAfter {
 }
 
 /// A structure for holding the response of a timing request.
-#[derive(Debug)]
```
Comment (dfoxfranke) [inline clock-bound-c/src/lib.rs]: Why remove this? Everything which can implement `Debug`, should. Likewise `Copy` and `Clone`.

**Review by dfoxfranke — CHANGES_REQUESTED** [review]

File: clock-bound-c/src/lib.rs (hunk 3 above)
Comment (basilnsindhu) [reply clock-bound-c/src/lib.rs]: Can add it back, I had it there for testing but then removed it

File: clock-bound-c/src/lib.rs (hunk 1 above)
Comment (basilnsindhu) [reply clock-bound-c/src/lib.rs]: Fixed. Heres a sample output 
```
Earliest start time for the timing request: "2022-02-23 17:47:43.459195163"
Latest finish time for the timing request: "2022-02-23 17:47:43.480670837"
Minimum execution duration of timing request: 89.999µs
Maximum execution duration of timing request: 90µs
```

File: clock-bound-c/src/lib.rs (hunk 2 above)
Comment (basilnsindhu) [reply clock-bound-c/src/lib.rs]: Fixed

File: clock-bound-c/src/lib.rs [hunk 4]
```diff
@@ -339,6 +359,71 @@ impl ClockBoundClient {
...
+        match self.socket.send(&mut request) {
+            Err(e) => return Err((ClockBoundCError::SendMessageError(e), Err(f))),
+            _ => {}
+        }
+        let mut response: [u8; 20] = [0; 20];
+        match self.socket.recv(&mut response) {
+            Err(e) => return Err((ClockBoundCError::ReceiveMessageError(e), Err(f))),
+            _ => {}
+        }
+        let response_earliest_start = NetworkEndian::read_u64(&response[4..12]);
```
[the remark this answers is not in the document]
Comment (basilnsindhu) [reply clock-bound-c/src/lib.rs]: I got the approach a little mixed up, fixed these

**Review by dfoxfranke — APPROVED** [review]

**Review by jacoblwisniewski — APPROVED** [review]

**Review by jacoblwisniewski — APPROVED** [review]

**Pull request merged**